src/ipset/ipset_hash_ipportnet.c

```diff
@@ -0,0 +1,245 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+#include <netinet/ip.h>
+#include <netinet/ip6.h>
+#include "ipset/ipset_hash.h"
+#include "ipset/pfxlen.h"
+
+typedef struct hash_ipportnet_elem {
+    union inet_addr ip;
+    uint8_t cidr;
+    union inet_addr ip2;
+    uint8_t proto;
+    uint16_t port;
+
+    /* data not evolved in hash calculation */
+    char comment[IPSET_MAXCOMLEN];
+    bool nomatch;
+} elem_t;
+
+static int
+hash_ipportnet_data_equal(const void *elem1, const void *elem2)
+{
+    elem_t *e2 = (elem_t *)elem2;
+
+    if (memcmp(elem1, elem2, offsetof(elem_t, comment)))
+        return COMPARE_INEQUAL;
+    if (e2->nomatch)
+        return COMPARE_EQUAL_REJECT;
+    return COMPARE_EQUAL_ACCEPT;
+}
+
+static void
+hash_ipportnet_do_list(struct ipset_member *member, void *elem, bool comment)
+{
+    elem_t *e = (elem_t *)elem;
+
+    member->addr = e->ip;
+    member->cidr = e->cidr;
+    member->port = ntohs(e->port);
+    member->proto = e->proto;
+    member->nomatch = e->nomatch;
+    member->addr2 = e->ip2;
+    
+    if (comment)
+        rte_strlcpy(member->comment, e->comment, IPSET_MAXCOMLEN);
+}
+
+static uint32_t
+hash_ipportnet_hashkey4(void *data, int len, uint32_t mask)
+{
+    elem_t *e = (elem_t *)data;
+
+    return (e->ip.in.s_addr * 31 + (((uint32_t)e->port << 16) |
+                (((uint32_t)e->proto) << 8) | (uint32_t)e->cidr)) & mask;
+}
+
+static int
+hash_ipportnet_adt4(int op, struct ipset *set, struct ipset_param *param)
+{
+    elem_t e;
+    int ret;
+    uint16_t port;
+    uint32_t ip, ip_to, ip2, ip2_to, ip2_from; 
+    ipset_adtfn adtfn = set->type->adtfn[op];
+
+    if (set->family != param->option.family)
+        return EDPVS_INVAL;
+
+    memset(&e, 0, sizeof(e));
+
+    e.cidr = param->cidr;
+    e.proto = param->proto;
+
+    if (op == IPSET_OP_TEST) {
+        e.ip.in.s_addr = param->range.min_addr.in.s_addr;
+        e.ip2.in.s_addr = param->range2.min_addr.in.s_addr;
+        e.port = htons(param->range.min_port);
+        return adtfn(set, &e, 0);
+    }
+
+    if (op == IPSET_OP_ADD) {
+        if (param->option.add.nomatch)
+            e.nomatch = true;
+        if (set->comment)
+            rte_strlcpy(e.comment, param->comment, IPSET_MAXCOMLEN);
+    }
+
+    ip = ntohl(param->range.min_addr.in.s_addr);
+    if (e.cidr) {
+        ip_set_mask_from_to(ip, ip_to, e.cidr);
+    } else {
+        ip_to = ntohl(param->range.max_addr.in.s_addr);
+    }
+    
+    ip2_from = ntohl(param->range2.min_addr.in.s_addr);;
+    if (param->cidr2) {
+        ip_set_mask_from_to(ip2_from, ip2_to, param->cidr2);
+    } else {
+        ip2_to = ntohl(param->range2.max_addr.in.s_addr);
+    }
+
+    do {
+        e.ip.in.s_addr = htonl(ip);
+        ip = ip_set_range_to_cidr(ip, ip_to, &e.cidr);
+        for (ip2 = ip2_from; ip2 <= ip2_to; ip2++) {
```

Comment:
Should ip and ip2 swap position to avoid confusing? Because this method is called IP-Port-Net.

Reply:
It's confined by the basic `hash_elem` memory arrangement. If a cidr not used, it should be cidr2.

```
typedef struct hash_elem {
    union inet_addr ip1;
    uint8_t cidr;
    union inet_addr ip2;
    uint8_t cidr2;
} elem_t;
```